File: frontend/src/types/sheet.ts
import type { Polygon } from "geojson";

/**
 * 도엽 (MapSheet) — 처리의 1차 단위.
 * PROJECT_BRIEF.md §3.1 기준. compare_type 은 PROMPTS §1 결정에 따라 'image-image' 단일.
 */
export type ReviewStatus =
  | "pending" // 미처리
  | "in_progress" // 진행중
  | "completed" // 완료
  | "on_hold"; // 보류

export type ObjectCategory = "building" | "road";

export type CompareType = "image-image";

export interface MapSheet {
  code: string; // 도엽코드 8자리
  name: string;
  region: string;
  bbox: [number, number, number, number]; // [minLon, minLat, maxLon, maxLat] EPSG:4326
  geometry: Polygon;
  area_km2: number;

  // 처리 상태
  review_status: ReviewStatus;
  reviewer: string | null;
  reviewed_at: string | null;

  // 작업 메타
  task_id: string;
  models: ObjectCategory[];
  compare_type: CompareType;
  standard_resource_id: number;
  compare_resource_id: number;

  // 메트릭
  f1_score: number | null;
  precision: number | null;
  recall: number | null;

  // 통계
  total_detections: number;
  reviewed_detections: number;
  tp_count: number;
  fp_count: number;
  fn_count: number;
}

/**
 * 도엽 목록 화면(/sheets)의 단일 진실 원천 필터.
 * 좌 사이드바·카드 리스트·지도 격자 색이 모두 동시 구독.
 */
export interface SheetFilter {
  search: string; // 도엽코드·도엽명
  reviewStatuses: ReviewStatus[]; // 다중선택 (빈 배열 = 전체)
  region: string | null; // 단일 권역 ('수도권북부', ...)
  categories: ObjectCategory[]; // 다중선택 (빈 배열 = 전체)
}

File: backend/app/schemas/sheet.py
"""도엽 (MapSheet) 스키마 — frontend src/types/sheet.ts 1:1."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.common import BboxArray, GeoJsonPolygon

ReviewStatus = Literal["pending", "in_progress", "completed", "on_hold"]
CompareType = Literal["image-image"]
ObjectCategory = Literal["building", "road"]


class MapSheet(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    name: str
    region: str
    bbox: BboxArray
    geometry: GeoJsonPolygon
    area_km2: float

    review_status: ReviewStatus
    reviewer: str | None = None
    reviewed_at: datetime | None = None

    # 그리드 도엽 (task 미할당) 은 모두 NULL.
    task_id: str | None = None
    models: list[ObjectCategory] = []
    compare_type: CompareType | None = None
    standard_resource_id: int | None = None
    compare_resource_id: int | None = None

    f1_score: float | None = None
    precision: float | None = None
    recall: float | None = None

    total_detections: int = 0


class SheetStatusUpdate(BaseModel):
    """PATCH /sheets/{code}/status 페이로드."""

    status: ReviewStatus
